src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+        aggfuncs: Aggregation functions for columns not in `by`.
+        errors: Error handling method.
+
+            - 'raise': Stop at the first error.
+            - 'coerce': Silently replace errors with `np.nan`.
+            - 'report': Replace errors with `np.nan` and return an error report.
+
+        errorfunc: A function with signature `f(x, e) -> Any`,
+            where `x` is the original input and `e` is the original error,
+            used when `errors='report'`.
+            By default, the original error is returned.
+
+    Returns:
+        Aggregated dataframe with `by` columns set as the index and
+        (if `errors`='report') an error report with a :class:`pd.Series` of errors
+        (or the value returned by `errorfunc`) for each column where errors occured.
+
+    Examples:
+        >>> import pandas as pd
+        >>> df = pd.DataFrame({
+        ...     'x': [0, 0, 1, 1],
+        ...     'y': pd.Series([2, 2, 2, 3], dtype='Int64')
+        ... })
+        >>> df.index = [0, 0, 0, 1]
+        >>> base = dict(df=df, by='x', aggfuncs={'y': unique})
+        >>> groupby_aggregate(**base, errors='raise')
+        Traceback (most recent call last):
+          ...
+        ValueError: Could not aggregate y: Not unique.
+        >>> groupby_aggregate(**base, errors='coerce')
+              y
+        x ...
+        0     2
+        1  <NA>
+        >>> result, report = groupby_aggregate(**base, errors='report')
+        >>> result
+              y
+        x ...
+        0     2
+        1  <NA>
+        >>> report['y']
+        x
+        1    Not unique.
+        Name: y, dtype: object
+        >>> errorfunc = lambda x, e: as_dict(x)
+        >>> result, report = groupby_aggregate(**base, errors='report', errorfunc=errorfunc)
+        >>> report['y']
+        x
+        1    {0: [2], 1: [3]}
+        Name: y, dtype: object
+    """
+    data_columns = [col for col in df.columns if col not in by]
+    dtypes = {col: df[col].dtype for col in data_columns}
+    if errors == "report":
+        reports = {}
+        # Prepare columns for error objects
+        df = df.astype({col: object for col in data_columns})
+    if errors == "raise":
+        aggfuncs = {
+            col: try_aggfunc(
+                func, method="raise", error=f"Could not aggregate {col}: {{e}}"
+            )
+            for col, func in aggfuncs.items()
+        }
+    elif errors == "coerce":
+        aggfuncs = {
+            col: try_aggfunc(func, method="coerce") for col, func in aggfuncs.items()
+        }
+    elif errors == "report":
+        aggfuncs = {
+            col: try_aggfunc(func, method="insert", error=errorfunc)
+            for col, func in aggfuncs.items()
+        }
+    if data_columns:
+        result = df.groupby(by).aggregate(aggfuncs)
+    else:
+        # groupby.aggregate drops index when aggfuncs is empty
+        result = df[by].drop_duplicates().set_index(by)
+    for col in data_columns:
```

Comment:
I don't understand how I would merge with the first `if errors == "report":` in the function, since that one needs to run before some intermediate code common to all `errors` scenarios. But I have merged it with the `if errors == "report":` immediately following it, and simplified the report building code somewhat.